securedrop_client/queue.py

```diff
@@ -0,0 +1,235 @@
+        '''
+        Checks if the file is valid.
+        :return: `True` if valid or unknown, `False` otherwise.
+        '''
+        if not etag:
+            logger.debug('No ETag. Skipping integrity check for file at {}'.format(file_path))
+            return True
+
+        alg, checksum = etag.split(':')
+
+        if alg == 'sha256':
+            hasher = hashlib.sha256()
+        else:
+            logger.debug('Unknown hash algorithm ({}). Skipping integrity check for file at {}'
+                         .format(alg, file_path))
+            return True
+
+        with open(file_path, 'rb') as f:
+            while True:
+                read_bytes = f.read(cls.CHUNK_SIZE)
+                if not read_bytes:
+                    break
+                hasher.update(read_bytes)
+
+        calculated_checksum = binascii.hexlify(hasher.digest()).decode('utf-8')
+        return calculated_checksum == checksum
+
+    def _decrypt_file(
+        self,
+        session: Session,
+        db_object: Union[File, Message],
+        file_path: str,
+    ) -> None:
+        file_uuid = db_object.uuid
+        server_filename = db_object.filename
+
+        # The filename contains the location where the file has been stored. On non-Qubes OSes, this
+        # will be the data directory. On Qubes OS, this will a ~/QubesIncoming directory. In case we
+        # are on Qubes, we should move the file to the data directory and name it the same as the
+        # server (e.g. spotless-tater-msg.gpg).
+        filepath_in_datadir = os.path.join(self.data_dir, server_filename)
+        shutil.move(file_path, filepath_in_datadir)
+        storage.mark_file_as_downloaded(file_uuid, session)
+
+        try:
+            self.gpg.decrypt_submission_or_reply(filepath_in_datadir, server_filename, is_doc=True)
+        except CryptoError as e:
+            logger.debug('Failed to decrypt file {}: {}'.format(server_filename, e))
+            storage.set_object_decryption_status_with_content(db_object, session, False)
+            raise e
+
+        storage.set_object_decryption_status_with_content(db_object, session, True)
+
+
+class RunnableQueue(QObject):
+
+    def __init__(self, api_client: API, session_maker: scoped_session) -> None:
+        super().__init__()
+        self.api_client = api_client
+        self.session_maker = session_maker
+        self.queue = Queue()  # type: Queue[ApiJob]
+        self.last_job = None  # type: Optional[ApiJob]
+
+    @pyqtSlot()
+    def process(self) -> None:  # pragma: nocover
+        self._process(False)
+
+    def _process(self, exit_loop: bool) -> None:
+        session = self.session_maker()
+        while True:
+            # retry the "cached" job if it exists, otherwise get the next job
+            if self.last_job is not None:
+                job = self.last_job
+                self.last_job = None
+            else:
+                job = self.queue.get(block=True)
+
+            try:
```

Comment:
I think it would be better to but a small for loop in `ApiJob._do_call_api` since that would bypass all the signal calling. I think we don't want to fire a failure signal for a job we're about to retry.